File: 139. Word Break/Solution.java
public class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        Set<String> wordSet = new HashSet<>(wordDict);
        int n = s.length();
        boolean[] dp = new boolean[n + 1];
        dp[0] = true; 
        for (int i = 1; i <= n; i++) {
            for (int j = 0; j < i; j++) {
                String currentWord = s.substring(j, i);
                if (dp[j] && wordSet.contains(currentWord)) {
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[n];
    }
}

File: 139. Word Break/Solution.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        wordSet = set(wordDict)
        n = len(s)
        dp = [False] * (n + 1)
        dp[0] = True
        for i in range(1, n + 1):
            for j in range(i):
                currentWord = s[j:i]
                if dp[j] and currentWord in wordSet:
                    dp[i] = True
                    break
        return dp[n]